src/privileged/sockets/api.js

```diff
@@ -0,0 +1,470 @@
+                  switch (meta[1]) {
+                    case 'j':
+                      string = JSON.parse(string);
+                    case 'n':
+                      gManager.dataListeners.forEach((listener) => {
+                        listener(id, string);
+                      });
+                      break;
+                    default:
+                      console.error("Unsupported serialization type (",meta[1],").");
+                      break;
+                  }
+                  inputStream.asyncWait(socketListener, 0, 0, tm.mainThread);
+                }
+              };
+              inputStream.asyncWait(socketListener, 0, 0, tm.mainThread);
+            }
+          });
+        },
+
+        onDataReceived: new ExtensionCommon.EventManager(
+          context,
+          "sockets.onDataReceived",
+          (fire) => {
+            let listener = (id, data) => {
+              fire.async(id, data);
+            };
+
+            gManager.dataListeners.add(listener);
+            return () => {
+              gManager.dataListeners.delete(listener);
+            };
+          }
+        ).api(),
+
+        async createSendingSocket() {
+          gManager.socketCount++;
+          gManager.sendingSocketMap.set(gManager.socketCount, {
+            stream: null,
+            bOutputStream: Cc["@mozilla.org/binaryoutputstream;1"]
+                             .createInstance(Ci.nsIBinaryOutputStream),
+          });
+          return gManager.socketCount;
+        },
+
+        connect(id, host, port) {
+          if (!gManager.sendingSocketMap.has(id)) {
+            return;
+          }
+          try {
+            let socket = gManager.sendingSocketMap.get(id);
+            var transport = socketService.createTransport(null, 0, host, port, null);
+            socket.stream = transport.openOutputStream(1, 4096, 1048575);
+            socket.bOutputStream.setOutputStream(socket.stream);
+            return true;
+          } catch (err) {
+            console.error(err,err.message);
+            return false;
+          }
+        },
+
+        sendData(id, data, serializationIdent) {
+          if (!gManager.sendingSocketMap.has(id)) {
+            return;
+          }
+          let socket = gManager.sendingSocketMap.get(id);
+          try {
+            let buff = bufferpack.pack('>Lc',[data.length, serializationIdent]);
+            socket.bOutputStream.writeByteArray(buff, buff.length);
+            socket.stream.write(data, data.length);
+            return true;
+          } catch (err) {
+            console.error(err,err.message);
+            return false;
+          }
+        },
+
+        close(id) {
+          if (!gManager.sendingSocketMap.has(id)) {
+            return;
+          }
+          gManager.sendingSocketMap.get(id).stream.close();
+        },
+      },
+    };
+  }
+};
+
+
+// The license that follows applies to all code below this line.
```

Comment:
Simply because otherwise, we need to get the moz-extension:// URL that locates the file and then manually load it into scope. I don't mind doing that though, I likely inlined it in the development process out of impatience.

Reply:
Keeping it as a separate dependency will probably make updates cleaner, but I'll defer to you on best practices here.

Reply:
If we were to have this as a separate dependency would it be easier to update automatically with the rest of the dependencies? Or do we still need to include the source in our repo? Either way, let's keep this as-is for this PR and file a follow-up bug if it seems useful.